### PR TITLE
Make message of ModuleEmitFailure error more detailed

### PR DESCRIPTION
**Pull request by tmat**
Adding more details to the error message allows for easier investigation of issues reported by customers. For example, https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1029186

### CODE REVIEW

**Comment by tmat** [comment]
@dotnet/roslyn-compiler PTAL

File: src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs [hunk 1]
```diff
@@ -2731,7 +2731,7 @@ internal override StrongNameKeys StrongNameKeys
                 if (moduleBeingBuilt.SourceModule.HasBadAttributes)
                 {
                     // If there were errors but no declaration diagnostics, explicitly add a "Failed to emit module" error.
-                    diagnostics.Add(ErrorCode.ERR_ModuleEmitFailure, NoLocation.Singleton, ((Cci.INamedEntity)moduleBeingBuilt).Name);
+                    diagnostics.Add(ErrorCode.ERR_ModuleEmitFailure, NoLocation.Singleton, ((Cci.INamedEntity)moduleBeingBuilt).Name, CodeAnalysisResources.ModuleHasInvalidAttributes);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs]: >CodeAnalysisResources.ModuleHasInvalidAttributes [](start = 134, length = 48)

This is likely to cause localization problems, I think we are using LocalizableResourceString in situations like this.

File: src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs [hunk 2]
```diff
@@ -181,7 +181,8 @@ private void SetGlobalErrorIfTrue(bool arg)
             // module or crashing.
             if (moduleBeingBuiltOpt != null && (methodCompiler._globalHasErrors || moduleBeingBuiltOpt.SourceModule.HasBadAttributes) && !diagnostics.HasAnyErrors() && !hasDeclarationErrors)
             {
-                diagnostics.Add(ErrorCode.ERR_ModuleEmitFailure, NoLocation.Singleton, ((Cci.INamedEntity)moduleBeingBuiltOpt).Name);
+                diagnostics.Add(ErrorCode.ERR_ModuleEmitFailure, NoLocation.Singleton, ((Cci.INamedEntity)moduleBeingBuiltOpt).Name,
+                    methodCompiler._globalHasErrors ? CodeAnalysisResources.UnableToDetermineSpecificCauseOfFailure : CodeAnalysisResources.ModuleHasInvalidAttributes);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Compiler/MethodCompiler.cs]: >CodeAnalysisResources.UnableToDetermineSpecificCauseOfFailure : CodeAnalysisResources.ModuleHasInvalidAttributes [](start = 54, length = 112)

This is likely to cause localization problems, I think we are using LocalizableResourceString in situations like this.

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs [hunk 3]
```diff
@@ -8920,26 +8919,23 @@ internal sealed class CSharpCompilerDiagnosticAnalyzer
             var emitResult2 = compilation2.Emit(peStream: new MemoryStream(), options: new EmitOptions(metadataOnly: true));
             Assert.False(emitResult2.Success);
             emitResult2.Diagnostics.Verify(
-    // error CS7038: Failed to emit module 'Test.dll'.
-    Diagnostic(ErrorCode.ERR_ModuleEmitFailure).WithArguments("Test.dll").WithLocation(1, 1)
-                );
+                // error CS7038: Failed to emit module 'Test.dll': Module has invalid attributes.
+                Diagnostic(ErrorCode.ERR_ModuleEmitFailure).WithArguments("Test.dll", CodeAnalysisResources.ModuleHasInvalidAttributes).WithLocation(1, 1));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs]: >CodeAnalysisResources.ModuleHasInvalidAttributes [](start = 86, length = 48)

This is likely going to cause a problem for Spanish_Unit_Tests once the strings are localized.

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFieldInitTests.cs [hunk 4]
```diff
@@ -689,8 +690,8 @@ public void FieldInitializerWithBadConstantValueDifferentModule()
             CreateCompilation(source2, new[] { new CSharpCompilationReference(compilation1) }, assemblyName: "2110a705-cc34-430b-9450-ca37031aa828")
                 .Emit(new System.IO.MemoryStream()).Diagnostics
                     .Verify(
-                    // error CS7038: Failed to emit module '2110a705-cc34-430b-9450-ca37031aa828'.
-                    Diagnostic(ErrorCode.ERR_ModuleEmitFailure).WithArguments("2110a705-cc34-430b-9450-ca37031aa828"));
+                    // error CS7038: Failed to emit module '2110a705-cc34-430b-9450-ca37031aa828': Unable to determine specific cause of the failure.
+                    Diagnostic(ErrorCode.ERR_ModuleEmitFailure).WithArguments("2110a705-cc34-430b-9450-ca37031aa828", CodeAnalysisResources.UnableToDetermineSpecificCauseOfFailure));
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFieldInitTests.cs]: >CodeAnalysisResources.UnableToDetermineSpecificCauseOfFailure) [](start = 118, length = 62)

This is likely going to cause a problem for Spanish_Unit_Tests once the strings are localized.

File: src/Compilers/Core/Portable/xlf/CodeAnalysisResources.es.xlf [hunk 5]
```diff
@@ -12,6 +12,11 @@
         <target state="translated">Se debe especificar un nombre de lenguaje para esta opción.</target>
         <note />
       </trans-unit>
+      <trans-unit id="ChangingVersionOfAssemblyReferenceIsNotAllowedDuringDebugging">
+        <source>Changing the version of an assembly reference is not allowed during debugging: '{0}' changed version to '{1}'.</source>
+        <target state="new">Changing the version of an assembly reference is not allowed during debugging: '{0}' changed version to '{1}'.</target>
```
Comment (AlekseyTs) [inline src/Compilers/Core/Portable/xlf/CodeAnalysisResources.es.xlf]: >Changing the version of an assembly reference is not allowed during debugging: '{0}' changed version to '{1}'. [](start = 28, length = 110)

Consider temporarily mangling new strings and making sure Spanish_Unit_Tests pass.

File: src/Compilers/CSharp/Test/Emit/Emit/EmitErrorTests.cs [hunk 6]
```diff
@@ -234,8 +234,8 @@ private static void VerifyEmitDiagnostics(string source2, CSharpCompilation comp
 
                 result.Diagnostics.Verify(expectedDiagnostics.Concat(new[]
                 {
-                    // error CS7038: Failed to emit module 'Test'.
-                    Diagnostic(ErrorCode.ERR_ModuleEmitFailure).WithArguments(compilation2.AssemblyName)
+                    // error CS7038: Failed to emit module 'Test': Unable to determine specific cause of the failure.
+                    Diagnostic(ErrorCode.ERR_ModuleEmitFailure).WithArguments(compilation2.AssemblyName, CodeAnalysisResources.UnableToDetermineSpecificCauseOfFailure)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Emit/Emit/EmitErrorTests.cs]: >CodeAnalysisResources.UnableToDetermineSpecificCauseOfFailure [](start = 105, length = 61)

This is likely going to cause a problem for Spanish_Unit_Tests once the strings are localized.

File: src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb [hunk 7]
```diff
@@ -257,7 +257,8 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 If (compiler.GlobalHasErrors OrElse moduleBeingBuiltOpt.SourceModule.HasBadAttributes) AndAlso Not hasDeclarationErrors AndAlso Not diagnostics.HasAnyErrors Then
                     ' If there were errors but no diagnostics, explicitly add
                     ' a "Failed to emit module" error to prevent emitting.
-                    diagnostics.Add(ERRID.ERR_ModuleEmitFailure, NoLocation.Singleton, moduleBeingBuiltOpt.SourceModule.Name)
+                    diagnostics.Add(ERRID.ERR_ModuleEmitFailure, NoLocation.Singleton, moduleBeingBuiltOpt.SourceModule.Name,
+                        If(compiler.GlobalHasErrors, CodeAnalysisResources.UnableToDetermineSpecificCauseOfFailure, CodeAnalysisResources.ModuleHasInvalidAttributes))
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/MethodCompiler.vb]: >CodeAnalysisResources.UnableToDetermineSpecificCauseOfFailure, CodeAnalysisResources.ModuleHasInvalidAttributes [](start = 53, length = 111)

This is likely to cause localization problems, I think we are using LocalizableResourceString in situations like this.

File: src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb [hunk 8]
```diff
@@ -2308,7 +2308,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
 
                 If moduleBeingBuilt.SourceModule.HasBadAttributes Then
                     ' If there were errors but no declaration diagnostics, explicitly add a "Failed to emit module" error.
-                    diagnostics.Add(ERRID.ERR_ModuleEmitFailure, NoLocation.Singleton, moduleBeingBuilt.SourceModule.Name)
+                    diagnostics.Add(ERRID.ERR_ModuleEmitFailure, NoLocation.Singleton, moduleBeingBuilt.SourceModule.Name, CodeAnalysisResources.ModuleHasInvalidAttributes)
```
Comment (AlekseyTs) [inline src/Compilers/VisualBasic/Portable/Compilation/VisualBasicCompilation.vb]: >CodeAnalysisResources.ModuleHasInvalidAttributes [](start = 123, length = 48)

This is likely to cause localization problems, I think we are using LocalizableResourceString in situations like this.

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 2)

File: src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit/Attributes/AttributeTests.cs]: > This is likely going to cause a problem for Spanish_Unit_Tests once the strings are localized.

The string should always be pulled from english resources, the test infrastructure always formats arguments this way. 

---
In reply to: [356197287](https://github.com/dotnet/roslyn/pull/40180#discussion_r356197287) [](ancestors = 356197287)

**Comment by AlekseyTs** [comment]
Done with review pass (iteration 4)

**Comment by tmat** [comment]
@cston @agocke PTAL, forgot I need 2 compiler reviews.

**Comment by cston** [comment]
LGTM

**Pull request merged**